plugins/modules/vault_kv2_write.py

```diff
@@ -0,0 +1,224 @@
+  data:
+    type: dict
+    required: true
+    description:
+      - KVv2 secret data to write/patch.
+  patch:
+    type: bool
+    default: false
+    description:
+      - Update an existing KVv2 secret with C(data) instead of overwriting.
+      - The secret must already exist.
+      - C(cas) must be C(false).
+  cas:
+    type: int
+    description:
+      - Perform a check-and-set operation.
+      - C(patch) must be C(false).
+"""
+
+EXAMPLES = r"""
+- name: Write/create a secret
+  community.hashi_vault.vault_kv2_write:
+    url: https://vault:8201
+    path: hello
+    data:
+      foo: bar
+
+- name: Patch an existing secret
+  community.hashi_vault.vault_kv2_write:
+    url: https://vault:8201
+    path: hello
+    data:
+      my: new key
+
+- name: Write with check and set
+  community.hashi_vault.vault_kv2_write:
+    url: https://vault:8201
+    path: caspath
+    cas: 0
+    data:
+      foo: bar
+"""
+
+RETURN = r"""
+raw:
+  type: dict
+  description: The raw Vault response.
+  returned: changed
+  sample:
+    auth:
+    data:
+      created_time: "2023-02-21T19:51:50.801757862Z"
+      custom_metadata:
+      deletion_time: ""
+      destroyed: false
+      version: 1
+    lease_duration: 0
+    lease_id: ""
+    renewable: false
+    request_id: 52eb1aa7-5a38-9a02-9246-efc5bf9581ec
+    warnings: null
+    wrap_info: null
+"""
+
+import traceback
+
+from ansible.module_utils._text import to_native
+from ansible.module_utils.basic import missing_required_lib
+from ansible_collections.community.hashi_vault.plugins.module_utils._hashi_vault_common import \
+    HashiVaultValueError
+from ansible_collections.community.hashi_vault.plugins.module_utils._hashi_vault_module import \
+    HashiVaultModule
+
+try:
+    import hvac
+except ImportError:
+    HAS_HVAC = False
+    HVAC_IMPORT_ERROR = traceback.format_exc()
+else:
+    HVAC_IMPORT_ERROR = None
+    HAS_HVAC = True
+
+
+def run_module():
+    argspec = HashiVaultModule.generate_argspec(
+        engine_mount_point=dict(type="str", default="secret"),
+        path=dict(type="str", required=True),
+        data=dict(type="dict", required=True),
+        cas=dict(type="int"),
+        patch=dict(type="bool", default=False),
+    )
+
+    module = HashiVaultModule(
+        argument_spec=argspec,
+        supports_check_mode=True,
+    )
+
+    if not HAS_HVAC:
+        module.fail_json(
+            msg=missing_required_lib("hvac"),
+            exception=HVAC_IMPORT_ERROR,
+        )
+
+    mount_point = module.params.get("engine_mount_point")
+    path = module.params.get("path")
+    cas = module.params.get("cas")
+    data = module.params.get("data")
+    patch = module.params.get("patch")
+
+    module.connection_options.process_connection_options()
+    client_args = module.connection_options.get_hvac_connection_options()
+    client = module.helper.get_vault_client(**client_args)
+
+    if cas and patch is True:
```

Comment:
```suggestion
    if cas is not None and patch is True:
```
A CAS value of 0 is valid, to perform a write only when the secret doesn't exist, so we don't want to allow `cas=0` to be used with `patch=True` presumably.

Let's also add tests for this condition.